sled-agent/src/services.rs

```diff
@@ -709,6 +828,71 @@ impl ServiceManager {
 
         Ok(())
     }
+
+    pub async fn ensure_switch(&self, request: Option<ServiceZoneRequest>) {
+        let log = &self.inner.log;
+        let mut switch_zone = self.inner.switch_zone.lock().await;
+
+        match (&*switch_zone, request) {
+            (SwitchZone::Disabled, Some(request)) => {
+                info!(log, "Enabling switch zone (new)");
+                let mgr = self.clone();
+                *switch_zone = SwitchZone::Initializing {
+                    request,
+                    initializer: tokio::task::spawn(async move {
+                        mgr.initialize_switch_zone().await
+                    }),
+                };
+            }
+            (SwitchZone::Initializing { .. }, Some(_)) => {
+                info!(log, "Enabling switch zone (already underway)");
+            }
+            (SwitchZone::Running(_), Some(_)) => {
+                info!(log, "Enabling switch zone (already complete)");
+            }
+            (SwitchZone::Disabled, None) => {
+                info!(log, "Disabling switch zone (already complete)");
+            }
+            (SwitchZone::Initializing { initializer, .. }, None) => {
+                info!(log, "Disabling switch zone (was initializing)");
+                initializer.abort();
```

Comment:
Does this introduce a chance for the switch zone to be orphaned mid-startup? If we're inside the `self.initialize_zone(&request).await` point inside `initialize_switch_zone` and we abort the task, will the zone only be torn down if we've made it far enough to have instantiated a `RunningZone` that will perform cleanup when it's dropped?

Reply:
Both this `ensure_switch` function and the `initialize_switch_zone` task perform operations on the "zone" while holding onto the `self.inner.switch_zone` lock, so the zone should "fully exist (in `SwitchZone::Running`) or not exist" at any point that we try to abort it.

That was my intent at least; does that look right top you?

Reply:
I was debating using https://docs.rs/tokio/latest/tokio/sync/oneshot/index.html when implementing this, to avoid any concerns of cancellation safety, if this seems too perilous to you - I could go down that route if this seems unsafe?

Reply:
Ah, I missed that we were holding the lock the whole time zone initialization is running. I think this looks right then, but just to make sure my understanding matches yours: The only ways we can get to this branch to abort the initializer is if either

a) someone calls `ensure_switch(Some(_))` and then immediately calls `ensure_switch(None)`, and the second `ensure_switch` happens to acquire the lock before the `initialize_switch_zone()` spawned by the first call can

b) zone initialization fails, in which case `initialize_switch_zone()` sleeps for 1 second before retrying (so any call to `ensure_switch(None)` during that sleep would hit this branch)

In either case, the abort is only cancelling a task that isn't doing anything meaningful (it either hasn't started yet or is sleeping between retries).

Reply:
I think what's here is safe because of the lock, but I think it would be clearer with message passing. Holding locks across `.await` points is somewhat confusing to me personally. Most of this is just personal preference though.

Reply:
I just added https://github.com/oxidecomputer/omicron/pull/1933/commits/282baee453752275a1ed70b58fb1e7c6cb78fa3c - I believe this is functionally the same, but it operates by asking the worker task to terminate, rather than invoking `abort`.

Does this seem preferable to y'all?

Reply:
I do like that better, but maybe just because I'm wary of cancellation issues in general? If others liked the abort version better I'd be fine going back to it.